File: opcua-webservice-node/examples/demoserver/src/models/company/skill/SkillClassType.ts
import { StatusCodes } from "opcua-webapi";
import {
    UaError,
    UaEUInformation,
    UaLocalizedText,
    UaStatusCode,
    UaValueRank,
    UaVariant,
} from "opcua-webapi-ts";
import {
    AddRequest,
    AddResponse,
    DeleteRequest,
    DeleteResponse,
    ElementCollectionType,
    GetDescriptorRequest,
    GetDescriptorResponse,
    GetPermissionRequest,
    GetPermissionResponse,
    ModifyAttributeRequest,
    ModifyAttributeResponse,
    ReadPropertyValuesRequest,
    ReadPropertyValuesResponse,
    UaDataTypes,
    UaVariable,
    UaVariableTypes,
    WritePropertyValuesRequest,
    WritePropertyValuesResponse
} from "opcua-webservice-node";

import { prisma } from "../../../connectors/prismaClient";
import { EmployeeTwinSpace } from "../../EmployeeTwinSpace";
import { SkillCategoryEnumType } from "./SkillCategoryEnumType";
import { SkillLevelEnumType } from "./SkillLevelEnumType";

export class SkillClassType extends ElementCollectionType {
    private readonly level: UaVariable;
    private readonly category: UaVariable;
    private readonly experience: UaVariable;

    constructor(
            skillLevelEnumType: SkillLevelEnumType,
            skillCategoryEnumType: SkillCategoryEnumType,
            twinSpace: EmployeeTwinSpace) {
        super("SkillClassType", new UaLocalizedText("SkillClass"), twinSpace);

        // Add start time data field
        this.level = this.addPropertyElement(
                "Level",
                new UaLocalizedText("Level"),
                new UaLocalizedText("The skill level of person"),
                skillLevelEnumType,
                true);

        // Add end time data field
        this.category = this.addPropertyElement(
                "Category",
                new UaLocalizedText("Category"),
                new UaLocalizedText("The category of skill"),
                skillCategoryEnumType,
                true);

        // Add experience data field
        this.experience = this.addPropertyElement(
                "Experience",
                new UaLocalizedText("Experience"),
                new UaLocalizedText("The year of experience required to be qualified"),
                UaDataTypes.Double,
                true,
                false, 
                UaValueRank.Scalar,     
                true,
                UaVariableTypes.BaseAnalogItemType);
        
        this.addSubElementOfProperty(
                this.experience,
                "EngineeringUnits",
                UaVariant.extensionObject(
                new UaEUInformation(
                    1,
                    new UaLocalizedText("Y"),
                    new UaLocalizedText("Year"),
                    null,
                ).toExtensionObject())
            );
    }
        
    override async onGetDescriptor(request: GetDescriptorRequest) : Promise<GetDescriptorResponse> {
        // Get department id from request
        const skillId = parseInt(request.id);

        if (Number.isNaN(skillId)) throw UaError.from(StatusCodes.BadNodeIdUnknown);

        // Database query
        let skill = await prisma.skill.findUnique({
            where: {
                ID: skillId
            }
        });

        if (skill === null) throw UaError.from(StatusCodes.BadNodeIdUnknown); 

        // Return department general information
        const response = new GetDescriptorResponse(
                new UaLocalizedText(skill.SkillName),
                new UaLocalizedText(skill.Description));

        return Promise.resolve(response);
    }
        
    override async onReadPropertyValues(request: ReadPropertyValuesRequest) : Promise<ReadPropertyValuesResponse> {
            // Get skill id from request
            const skillId = parseInt(request.id);

            // Database query
            const skill = await prisma.skill.findUnique({
                where: { ID: skillId },
                select: {
                    Level: true,
                    Category: true,
                    YearOfExperience: true
                }
            });
            if (skill === null) throw UaError.from(StatusCodes.BadNodeIdUnknown);

            // Return field data
            const response = new ReadPropertyValuesResponse();
            response.setValue(this.level.name, UaVariant.integer(EmployeeTwinSpace.SkillLevelEnumType.parse(skill.Level)));
            response.setValue(this.category.name, UaVariant.integer(EmployeeTwinSpace.SkillCategoryEnumType.parse(skill.Category)));
            response.setValue(this.experience.name, UaVariant.integer(skill.YearOfExperience));

            return response;
    }
    
    override async onWritePropertyValues(request: WritePropertyValuesRequest): Promise<WritePropertyValuesResponse>
    {
        const skillId = parseInt(request.id);
        if (Number.isNaN(skillId)) throw UaError.from(StatusCodes.BadNodeIdUnknown);

        let response = new WritePropertyValuesResponse();
        
        let levelValue = request.propertyNamesAndValues.get(this.level.name)?.toNumber();
        let categoryValue = request.propertyNamesAndValues.get(this.category.name)?.toNumber();
        let experienceValue = request.propertyNamesAndValues.get(this.experience.name)?.toNumber();
        
        let level = undefined;
        let category = undefined;
        let experience = undefined;
        
        if (levelValue !== null && levelValue !== undefined)
        {            
            level = SkillLevelEnumType.fromNumber(levelValue);
        }
            
        if (categoryValue !== null && categoryValue !== undefined)
        {
            
            category = SkillCategoryEnumType.fromNumber(categoryValue);
        }

        if (experienceValue !== null && experienceValue !== undefined)
        {
            if (experienceValue >= 0 && experienceValue <= 50) 
            {
                experience = experienceValue;
            } else {
                response.setWriteValueResult(this.experience.name, UaStatusCode.from(StatusCodes.BadInvalidArgument));
            }
        }

        if (level === undefined && category === undefined && experience === undefined)
        {
            throw UaError.from(StatusCodes.BadInvalidArgument);
        }

        try {
            await prisma.skill.update({
                where: { ID: skillId },
                data: {
                    Level: level,
                    Category: category,
                    YearOfExperience: experience }
                });
        } catch (error) {
            throw UaError.from(StatusCodes.BadDataUnavailable);
        }

        return response;
    } 

    override async onGetPermission(request: GetPermissionRequest): Promise<GetPermissionResponse>
    {
        return new GetPermissionResponse(true, true, true);
    }

    override async onAddElement(request: AddRequest): Promise<AddResponse>
    {        
        if (!request.displayName.text) throw UaError.from(StatusCodes.BadInvalidArgument);
        
        let skill = await prisma.skill.create({
            data: {
                SkillName: request.displayName.text,
                Description: null,
                Level: 3,
                Category: "",
                YearOfExperience: 0
            }
        });

        return new AddResponse(skill.ID.toString());
    }

    override async onDeleteElement(request: DeleteRequest): Promise<DeleteResponse> {        
        const skillId = parseInt(request.id);
        if (Number.isNaN(skillId)) throw UaError.from(StatusCodes.BadNodeIdUnknown);

        let statusCode = UaStatusCode.from(StatusCodes.Good);
        try {
            await prisma.skill.delete({
                where: { ID: skillId }
            });
        } catch (error) {
            statusCode = UaStatusCode.from(StatusCodes.BadNodeIdUnknown);
        }

        return new DeleteResponse(statusCode);
    }

    override async onRename(request: ModifyAttributeRequest): Promise<ModifyAttributeResponse>
    {
        if (!request.text.text) throw UaError.from(StatusCodes.BadInvalidArgument);

        const skillId = parseInt(request.id);
        if (Number.isNaN(skillId)) throw UaError.from(StatusCodes.BadNodeIdUnknown);

        let statusCode = UaStatusCode.from(StatusCodes.Good);
        try {
            await prisma.skill.update({
                where: { ID: skillId },
                data: {
                    SkillName: request.text.text
                }
            });
        } catch (error) {
            statusCode = UaStatusCode.from(StatusCodes.BadNodeIdUnknown);
        }

        return new ModifyAttributeResponse(statusCode);
    }

    override async onSetDescriptor(request: ModifyAttributeRequest): Promise<ModifyAttributeResponse>
    {
        console.log("onSetDescriptor");
        if (!request.text.text) throw UaError.from(StatusCodes.BadInvalidArgument);

        const skillId = parseInt(request.id);
        if (Number.isNaN(skillId)) throw UaError.from(StatusCodes.BadNodeIdUnknown);

        let statusCode = UaStatusCode.from(StatusCodes.Good);
        try {
            await prisma.skill.update({
                where: { ID: skillId },
                data: {
                    Description: request.text.text
                }
            });
        } catch (error) {
            statusCode = UaStatusCode.from(StatusCodes.BadNodeIdUnknown);
        }

        return new ModifyAttributeResponse(statusCode);
    }
}

File: opcua-webservice-spring/examples/uawebservicedemo/src/main/java/org/opcfoundation/uawebservicedemo/models/company/skillclassification/SkillClassType.java
package org.opcfoundation.uawebservicedemo.models.company.skillclassification;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.DateTime;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.structured.EUInformation;
import org.opcfoundation.uawebservicedemo.database.EmployeeDatabase;
import org.opcfoundation.uawebservicedemo.database.entity.EmployeeData;
import org.opcfoundation.uawebservicedemo.database.entity.Skill;
import org.opcfoundation.uawebservicedemo.models.company.skillclassification.enumeration.SkillCategory;
import org.opcfoundation.uawebservicedemo.models.company.skillclassification.enumeration.SkillLevel;
import org.opcfoundation.uawebservicedemo.database.mapper.SkillMapper;
import org.opcfoundation.uawebservicedemo.models.EmployeeTwinSpace;
import org.opcfoundation.uawebservicedemo.models.company.skillclassification.enumeration.SkillCategoryEnumType;
import org.opcfoundation.uawebservicedemo.models.company.skillclassification.enumeration.SkillLevelEnumType;
import org.opcfoundation.uawebservicedemo.models.employee.personaldata.enumeration.Sex;
import org.opcfoundation.webserver.addressspace.nodes.UaVariable;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaDataTypes;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaVariableTypes;
import org.opcfoundation.webserver.digitaltwin.element.ElementCollectionType;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDescriptorRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDescriptorResponse;
import org.opcfoundation.webserver.service.message.digitaltwin.ReadPropertyValuesRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.ReadPropertyValuesResponse;
import org.opcfoundation.webserver.types.common.UaStructureUtilities;

import java.time.ZoneId;
import java.util.Set;
import java.util.concurrent.CompletableFuture;

public class SkillClassType extends ElementCollectionType {
    private final UaVariable level;
    private final UaVariable category;
    private final UaVariable experience;

    public SkillClassType(
            SkillLevelEnumType skillLevelEnumType,
            SkillCategoryEnumType skillCategoryEnumType,
            EmployeeTwinSpace twinSpace) {
        super("SkillClassType", new LocalizedText("SkillClass"), twinSpace);

        // Add start time data field
        level = addPropertyElement(
                "Level",
                new LocalizedText("Level"),
                new LocalizedText("The skill level of person"),
                skillLevelEnumType,
                false);

        // Add end time data field
        category = addPropertyElement(
                "Category",
                new LocalizedText("Category"),
                new LocalizedText("The category of skill"),
                skillCategoryEnumType,
                false);

        // Add experience data field
        experience = addPropertyElement(
                "Experience",
                new LocalizedText("Experience"),
                new LocalizedText("The year of experience required to be qualified"),
                UaDataTypes.Double,
                false,
                false,
                null,
                UaVariableTypes.BaseAnalogType,
                true);

        addSubElementOfProperty(
                experience,
                "EngineeringUnits",
                UaStructureUtilities.toVariant(new EUInformation(null,-1, new LocalizedText("Y"), new LocalizedText("Year"))));
    }

    // Return general information such as name, description
    @Override
    public CompletableFuture<GetDescriptorResponse> onGetDescriptor(GetDescriptorRequest request) {
        try
        {
            // Get department id from request
            Integer skillId = Integer.parseInt(request.getId());

            // Database query
            SkillMapper skillMapper = EmployeeDatabase.getSkillMapper();
            Skill skill = skillMapper.getSkill(skillId);
            if (null == skill) throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);

            // Return department general information
            GetDescriptorResponse response = new GetDescriptorResponse(
                    new LocalizedText(skill.getName()), // Object name
                    new LocalizedText(skill.getDescription())); // Object description

            return CompletableFuture.completedFuture(response);

        } catch (Exception e) {
            throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);
        }
    }

    @Override
    public CompletableFuture<ReadPropertyValuesResponse> onReadPropertyValues(ReadPropertyValuesRequest request) {
        try
        {
            // Get skill id from request
            Integer skillId = Integer.parseInt(request.getId());

            // Database query
            SkillMapper skillMapper = EmployeeDatabase.getSkillMapper();
            Skill skill = skillMapper.getSkillData(skillId);
            if (null == skill) throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);

            // Return field data
            ReadPropertyValuesResponse response = new ReadPropertyValuesResponse();

            SkillLevel levelValue = SkillLevel.fromInt(skill.getLevel());
            SkillCategory categoryValue = SkillCategory.fromString(skill.getCategory());

            response.setValue(level.name(), Variant.ofInt32(levelValue.getCode()));
            response.setValue(category.name(), Variant.ofInt32(categoryValue.getCode()));
            response.setValue(experience.name(),Variant.ofInt32(skill.getYearOfExperience()));

            return CompletableFuture.completedFuture(response);
        } catch (Exception e) {
            throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);
        }
    }
}
